### PR TITLE
Unified OutputWriter and BufferWriter

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
cc: @ahsonkhan, @davidfowl, @joshfree 

I used the implementation of basic writer logic (fields, Advance, Ensure, etc.) from https://github.com/aspnet/benchmarks/tree/dev/src/PlatformBenchmarks

Added the ability to write ints and IWritables from the removed BufferWriter.

The perf results are super close (probably within the margin of error). Once the writer and Utf8String are in corefx, I think we will be able to switch the ProtformBenchmark implementation to use these.

 Benchmarks.dll                               | Metric               | Unit  | Iterations |    Average |    STDEV.S |        Min |        Max
:-------------------------------------------- |:-------------------- |:-----:|:----------:| ----------:| ----------:| ----------:| ----------:
 BufferWriterBench.BufferWriterCopyPlaintext  | Duration             | msec  |    100     |     49.239 |      3.057 |     44.772 |     66.541
 BufferWriterBench.BufferWriterCopyPlaintext  | Instructions Retired | count |    100     | 6.473E+008 | 1.190E+006 | 6.460E+008 | 6.550E+008
 BufferWriterBench.BufferWriterCopyPlaintext  | Cache Misses         | count |    100     |  81633.280 |  14440.590 |  49152.000 | 1.270E+005
 BufferWriterBench.BufferWriterPlaintext      | Duration             | msec  |    100     |     54.272 |      4.365 |     49.248 |     71.213
 BufferWriterBench.BufferWriterPlaintext      | Instructions Retired | count |    100     | 6.531E+008 | 1.972E+006 | 6.500E+008 | 6.620E+008
 BufferWriterBench.BufferWriterPlaintext      | Cache Misses         | count |    100     |  96788.480 |  26961.004 |  53248.000 | 1.720E+005
 BufferWriterBench.BufferWriterPlaintextUtf8  | Duration             | msec  |    100     |     52.471 |      1.981 |     49.635 |     60.774
 BufferWriterBench.BufferWriterPlaintextUtf8  | Instructions Retired | count |    100     | 6.534E+008 | 1.017E+006 | 6.520E+008 | 6.560E+008
 BufferWriterBench.BufferWriterPlaintextUtf8  | Cache Misses         | count |    100     |  86548.480 |  17207.879 |  57344.000 | 1.475E+005
 BufferWriterBench.PlatfromBenchmarkPlaintext | Duration             | msec  |    100     |     50.390 |      2.965 |     46.137 |     65.976
 BufferWriterBench.PlatfromBenchmarkPlaintext | Instructions Retired | count |    100     | 6.601E+008 | 1.348E+006 | 6.580E+008 | 6.650E+008
 BufferWriterBench.PlatfromBenchmarkPlaintext | Cache Misses         | count |    100     |  86425.600 |  21449.999 |  53248.000 | 1.516E+005

### CODE REVIEW

File: src/System.Buffers.ReaderWriter/System/Buffers/Writer/OutputWriter.cs [hunk 1]
```diff
@@ -6,31 +6,23 @@
...
-    public static class OutputWriter
-    {
-        public static OutputWriter<T> Create<T>(T output) where T : IBufferWriter<byte>
-        {
-            return new OutputWriter<T>(output);
-        }
-    }
-
-    public ref struct OutputWriter<T> where T : IBufferWriter<byte>
+    public ref partial struct BufferWriter<T> where T : IBufferWriter<byte>
```
Comment (YohDeadfall) [inline src/System.Buffers.ReaderWriter/System/Buffers/Writer/OutputWriter.cs]: Since now it's `BufferWriter<T>`, the file should be renamed.

File: src/System.Buffers.ReaderWriter/System/Buffers/Writer/OutputWriter.cs [hunk 2]
```diff
@@ -6,31 +6,23 @@
...
-        public OutputWriter(T output)
+        public BufferWriter(T output)
         {
             _buffered = 0;
             _output = output;
             _span = output.GetSpan();
         }
 
-        public Span<byte> Span => _span;
+        public Span<byte> Buffer => _span;
```
Comment (YohDeadfall) [inline src/System.Buffers.ReaderWriter/System/Buffers/Writer/OutputWriter.cs]: Why `Buffer`? It would be better to name the property `Span` like it done in [`Memory`](https://github.com/dotnet/coreclr/blob/8bad66fca4442839adea749e7a35b0dda8023eda/src/mscorlib/shared/System/Memory.cs#L186).

File: src/System.Buffers.ReaderWriter/System/Buffers/Writer/OutputWriter.cs (hunk 1 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.ReaderWriter/System/Buffers/Writer/OutputWriter.cs]: Yeah, I will rename it in a subsequent change, otherwise github gets often confused about the change and treats files are deleted/added.

File: src/System.Buffers.ReaderWriter/System/Buffers/Writer/OutputWriter.cs (hunk 2 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.ReaderWriter/System/Buffers/Writer/OutputWriter.cs]: I thought buffer communicates better that the write needs to be Flushed (and one of the fields is called _buffered).

File: src/System.Buffers.ReaderWriter/System/Buffers/Writer/OutputWriter_overloads.cs [hunk 3]
```diff
@@ -0,0 +1,114 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers.Text;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers.Writer
+{
+    public ref partial struct BufferWriter<T> where T : IBufferWriter<byte>
```
Comment (ahsonkhan) [inline src/System.Buffers.ReaderWriter/System/Buffers/Writer/OutputWriter_overloads.cs]: Why does this need to be generic when the T is constrained to only a single type?

File: src/System.Buffers.ReaderWriter/System/Buffers/Writer/OutputWriter_overloads.cs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            while (!Utf8Formatter.TryFormat(value, Buffer, out written, format))
+            {
+                Enlarge();
+            }
+            Advance(written);
+        }
+
+        public void Write(string value)
+        {
+            var utf16Bytes = value.AsSpan().AsBytes();
```
Comment (ahsonkhan) [inline src/System.Buffers.ReaderWriter/System/Buffers/Writer/OutputWriter_overloads.cs]: nit: remove use of var

File: tests/Benchmarks/BufferWriterBench.cs [hunk 5]
```diff
@@ -0,0 +1,527 @@
...
+using System.Buffers.Writer;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Text.Utf8;
+using System.Threading;
+
+public class BufferWriterBench
+{
+    private static AsciiString _crlf = "\r\n";
```
Comment (ahsonkhan) [inline tests/Benchmarks/BufferWriterBench.cs]: nit: static fields should be prefixed with `s_`

File: tests/Benchmarks/BufferWriterBench.cs [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+    private static Utf8String _http11OKU8 = (Utf8String)"HTTP/1.1 200 OK\r\n";
+    private static Utf8String _headerServerU8 = (Utf8String)"Server: Custom";
+    private static Utf8String _headerContentLengthU8 = (Utf8String)"Content-Length: ";
+    private static Utf8String _headerContentLengthZeroU8 = (Utf8String)"Content-Length: 0\r\n";
+    private static Utf8String _headerContentTypeTextU8 = (Utf8String)"Content-Type: text/plain\r\n";
+    private static Utf8String _plainTextBodyU8 = (Utf8String)"Hello, World!";
+
+    static Sink _sink = new Sink(4096);
+
+    const int InnerItterations = 1000000;
```
Comment (ahsonkhan) [inline tests/Benchmarks/BufferWriterBench.cs]: typo: `InnerItterations `

File: tests/Benchmarks/BufferWriterBench.cs [hunk 7]
```diff
@@ -0,0 +1,527 @@
...
+                if (Utf8Formatter.TryFormat(number, Span, out bytesWritten))
+                {
+                    Advance(bytesWritten);
+                }
+            }
+        }
+    }
+}
+
+// copy from System.Buffers.ReaderWriter to isolate cross-dll calls.
```
Comment (ahsonkhan) [inline tests/Benchmarks/BufferWriterBench.cs]: Why is this necessary?

File: src/System.Buffers.ReaderWriter/System/Buffers/Writer/OutputWriter_overloads.cs (hunk 3 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.ReaderWriter/System/Buffers/Writer/OutputWriter_overloads.cs]: So that IBufferWriter implementations can be structs and be written to without boxing.

File: tests/Benchmarks/BufferWriterBench.cs (hunk 7 above)
Comment (KrzysztofCwalina) [reply tests/Benchmarks/BufferWriterBench.cs]: For some reason source code compiled into the test is a bit faster. I wanted to have totally apples to apples comparison between BufferWriters and so I wanted to have them both in the same project.

File: tests/Benchmarks/BufferWriterBench.cs (hunk 6 above)
Comment (KrzysztofCwalina) [reply tests/Benchmarks/BufferWriterBench.cs]: thanks. will fix

File: tests/Benchmarks/BufferWriterBench.cs (hunk 5 above)
Comment (KrzysztofCwalina) [reply tests/Benchmarks/BufferWriterBench.cs]: wil do
